preview/clean.js

```diff
@@ -0,0 +1,181 @@
+'use strict';
+
+const child_process = require('child_process');
+const dedent = require('dedent');
+const https = require('https');
+const os = require('os');
+const path = require('path');
+
+module.exports = Cleaner;
+module.exports.exec_git = exec_git;
+
+function Cleaner(token, repo, cache_dir, tmp_dir) {
+  this.repo = repo;
```

Comment:
Doesn't really make a difference, but just wanted to note that you don't really need to use `this` with `new Cleaner`.

The function arguments are captured in the closure, so you could define the functions inside the body, and return an object at the end which would have the same effect. Possibly even better, move the functions out of the body, and modify their definitions so you call them with the data they need.


```Javascript
const clone = (localPath, cacheDir, repoName) =>
  exec_git([
      'clone',
      '--bare',
      '--reference', `${cacheDir}/${repoName}`,
      repo, localPath])

const getRepoName = repo => {
  let repo_name = path.basename(repo)
  if (!repo_name.endsWith('.git')) {
    repo_name += '.git';
  }
  return repoName
}

export const Cleaner = (token, repo, cache_dir, tmp_dir) => {
  const repoName = getRepoName(repo)
  const localPath = `${tmp_dir}/${repoName}`
  
  const showHeads = () => exec_git(['show-ref', '--heads'], {cwd: localPath})

  ;; snipped

  return {
    run: () =>
      clone(localPath, cache_dir, repoName)
        .then(show_heads)
        .then(heads_to_prs)
        .then(cleanup_closed_prs),
    showHeads: showHeads ;; just an example of exporting a function that's leveraging data int the closure.
    
  }
}
```

It could make things a little easier to test if you made the inner functions standalone while keeping the ease-of-use API the same.

This comment isn't an action item, just wanted to bring up this strategy.